lsps0/protocol_server.go

```diff
@@ -0,0 +1,55 @@
+package lsps0
+
+import "context"
+
+type ProtocolServer interface {
+	ListProtocols(
+		ctx context.Context,
+		req *ListProtocolsRequest,
+	) (*ListProtocolsResponse, error)
+}
+type protocolServer struct {
+	protocols []uint32
+}
+
+type ListProtocolsRequest struct{}
+type ListProtocolsResponse struct {
+	Protocols []uint32 `json:"protocols"`
+}
+
+func NewProtocolServer(supportedProtocols []uint32) ProtocolServer {
+	return &protocolServer{
+		protocols: supportedProtocols,
+	}
+}
+
+func (s *protocolServer) ListProtocols(
+	ctx context.Context,
+	req *ListProtocolsRequest,
+) (*ListProtocolsResponse, error) {
+	return &ListProtocolsResponse{
+		Protocols: s.protocols,
+	}, nil
+}
+
+func RegisterProtocolServer(s ServiceRegistrar, p ProtocolServer) {
+	s.RegisterService(
+		&ServiceDesc{
+			ServiceName: "lsps0",
+			HandlerType: (*ProtocolServer)(nil),
```

Comment:
Why do we need the `HandlerType` ? Where do we use it?

Reply:
> I took it from the grpc service desc. It was used to check the runtime type of the handler passed via `RegisterService`, whether it implements the interface in the service description. I had removed that check, because I misunderstood it. Reintroduced the check now. Let me know if you think that's a good idea.

I see. As long as we have generic RegisterService underneath then I agree the check is a good idea.
I wonder if we can simplify and add the RegisterXXXServer methods (e.g RegisterProtocolServer) directly as a method of the Server struct and have that type safety check as part of the method prototype. 
What do yo think?

Reply:
> I wonder if we can simplify and add the RegisterXXXServer methods (e.g RegisterProtocolServer) directly as a method of the Server struct and have that type safety check as part of the method prototype.

I don't see a way to do that without creating a circular reference between the implementing services and the main server, do you have something in mind?

Reply:
I see... I now understand that the interfaces are not declared in the base module (lsp0) but every addition introduces its own interface so indeed we can't do it the way I suggested.